File: reference-implementations/agt-go/internal/mcp/mcp.go
// Package mcp reads wrapped MCP messages, protocols/MCP/* (extend_mcp.md).
// A wrapped call carries one MCP JSON-RPC 2.0 message intact in
// params.payload: a request or notification the Observed Agent is about to
// send, or the response it received, which it wraps under the method the
// response answers.
package mcp

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strings"

	jsonv2 "github.com/go-json-experiment/json"
	"github.com/go-json-experiment/json/jsontext"
)

// Protocol is the name MCP carries in protocols/MCP/* and in
// wrapped_protocols.
const Protocol = "MCP"

// methodName is the form every MCP method name takes: slash-separated
// segments of letters, digits and underscores (initialize,
// tools/call, notifications/resources/list_changed).
var methodName = regexp.MustCompile(`^[A-Za-z0-9_]+(/[A-Za-z0-9_]+)*$`)

// Handler is the wrapped-protocol handler for MCP. It reads every MCP
// method, since it checks the JSON-RPC shape every MCP message shares and
// leaves the method's meaning to the policy engine; a list of methods would
// refuse the next MCP revision's.
type Handler struct{}

// Protocol implements method.WrappedHandler.
func (Handler) Protocol() string { return Protocol }

// Supports implements method.WrappedHandler.
func (Handler) Supports(method string) bool { return methodName.MatchString(method) }

// Message is the part of a wrapped MCP message the Guardian and the policy
// engine read. A request or notification has Method; a response has Result
// or Error.
type Message struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      jsontext.Value  `json:"id,omitzero"`
	Method  string          `json:"method,omitzero"`
	Params  json.RawMessage `json:"params,omitzero"`
	Result  json.RawMessage `json:"result,omitzero"`
	Error   json.RawMessage `json:"error,omitzero"`
}

// IsResponse reports whether m is a response rather than a request or a
// notification.
func (m Message) IsResponse() bool { return m.Method == "" }

// Read decodes payload as the wrapped MCP message of method and checks it
// against JSON-RPC 2.0: version "2.0"; a request's method equal to the
// wrapped method and its params, when present, an object; a response's id
// present and exactly one of result, an object, and error, an object with
// an integer code and a string message.
func Read(method string, payload json.RawMessage) (Message, error) {
	var m Message
	if err := jsonv2.Unmarshal(payload, &m); err != nil {
		return m, fmt.Errorf("the wrapped MCP message cannot be read: %w", err)
	}
	if m.JSONRPC != "2.0" {
		return m, errors.New(`the wrapped MCP message must have jsonrpc "2.0"`)
	}
	if len(m.ID) > 0 && m.ID.Kind() != '"' && m.ID.Kind() != '0' {
		return m, errors.New("the wrapped MCP message's id must be a string or a number")
	}
	if !m.IsResponse() {
		if m.Method != method {
			return m, fmt.Errorf("the wrapped MCP message is %s, and the envelope wraps %s", m.Method, method)
		}
		if len(m.Result) > 0 || len(m.Error) > 0 {
			return m, errors.New("a wrapped MCP request carries neither result nor error")
		}
		if len(m.Params) > 0 && !isObject(m.Params) {
			return m, errors.New("the wrapped MCP message's params must be an object")
		}
		return m, nil
	}
	if len(m.ID) == 0 {
		return m, errors.New("a wrapped MCP response must carry the id of the request it answers")
	}
	switch {
	case len(m.Result) > 0 && len(m.Error) > 0, len(m.Result) == 0 && len(m.Error) == 0:
		return m, errors.New("a wrapped MCP response carries exactly one of result and error")
	case len(m.Result) > 0 && !isObject(m.Result):
		return m, errors.New("a wrapped MCP response's result must be an object")
	case len(m.Error) > 0:
		var e struct {
			Code    *int64  `json:"code"`
			Message *string `json:"message"`
		}
		if err := jsonv2.Unmarshal(m.Error, &e); err != nil || e.Code == nil || e.Message == nil {
			return m, errors.New("a wrapped MCP response's error must be an object with an integer code and a string message")
		}
	}
	return m, nil
}

// Check implements method.WrappedHandler.
func (Handler) Check(method string, payload json.RawMessage) error {
	_, err := Read(method, payload)
	return err
}

// ParameterOverridePointer implements method.WrappedHandler.
func (Handler) ParameterOverridePointer(argument string) string {
	escaped := strings.NewReplacer("~", "~0", "/", "~1").Replace(argument)
	return "/params/arguments/" + escaped
}

func isObject(b json.RawMessage) bool {
	b = bytes.TrimSpace(b)
	return len(b) > 0 && b[0] == '{'
}

File: reference-implementations/agt-python/src/acs_guardian/mcp.py
"""The wrapped MCP handler: ``protocols/MCP/*`` (extend_mcp.md).

A wrapped call carries one MCP JSON-RPC 2.0 message intact in
``params.payload``: a request or notification the Observed Agent is about to
send, or the response it received, which it wraps under the method the
response answers. The message reaches the policy engine intact; this module
is the only protocol-specific part of a wrapped call — the envelope, the
session, the chain, the decision and the signature are the Guardian's,
shared with the native hooks.

``read`` implements the JSON-RPC 2.0 shape checks every MCP message shares;
the same checks the Go port contributed in the open PR #169
(``reference-implementations/agt-go/internal/mcp/mcp.go``) makes, so the two
refuse the same messages. The method name is not compared for a response: a
response carries no ``method``, and it is wrapped under the method of the
request it answers.
"""

from __future__ import annotations

import re
from typing import Any

from .errors import INVALID_PARAMS, AcsError

PROTOCOL = "MCP"

# Every MCP method name: slash-separated segments of letters, digits and
# underscores (initialize, tools/call, notifications/resources/list_changed).
_METHOD_NAME = re.compile(r"^[A-Za-z0-9_]+(/[A-Za-z0-9_]+)*$")


def supports(method: str) -> bool:
    """Whether the handler reads this MCP method (any well-formed name)."""
    return bool(_METHOD_NAME.match(method))


def read(method: str, payload: Any) -> dict[str, Any]:
    """Decode and check a wrapped MCP message, or raise INVALID_PARAMS (-32602).

    ``method`` is the part after ``protocols/MCP/``. A request's own method
    must equal it; a response must carry an id and exactly one of ``result``
    (an object) and ``error`` (an object with an integer code and a string
    message).
    """
    if not isinstance(payload, dict):
        raise AcsError(INVALID_PARAMS, "the wrapped MCP message must be an object", data={"method": method})
    if payload.get("jsonrpc") != "2.0":
        raise AcsError(
            INVALID_PARAMS, 'the wrapped MCP message must have jsonrpc "2.0"', data={"method": method}
        )
    has_id = "id" in payload
    if has_id:
        message_id = payload["id"]
        # A JSON null id is not a string or a number, and is reserved for
        # responses to unparseable requests, so it is refused here too.
        if isinstance(message_id, bool) or not isinstance(message_id, (str, int)):
            raise AcsError(
                INVALID_PARAMS, "the wrapped MCP message's id must be a string or a number", data={"method": method}
            )

    is_response = "method" not in payload
    if not is_response:
        if payload.get("method") != method:
            raise AcsError(
                INVALID_PARAMS,
                f"the wrapped MCP message is {payload.get('method')}, and the envelope wraps {method}",
                data={"method": method},
            )
        if "result" in payload or "error" in payload:
            raise AcsError(
                INVALID_PARAMS, "a wrapped MCP request carries neither result nor error", data={"method": method}
            )
        if "params" in payload and not isinstance(payload["params"], dict):
            raise AcsError(
                INVALID_PARAMS, "the wrapped MCP message's params must be an object", data={"method": method}
            )
        return payload

    if not has_id:
        raise AcsError(
            INVALID_PARAMS,
            "a wrapped MCP response must carry the id of the request it answers",
            data={"method": method},
        )
    has_result = "result" in payload
    has_error = "error" in payload
    if has_result == has_error:
        raise AcsError(
            INVALID_PARAMS,
            "a wrapped MCP response carries exactly one of result and error",
            data={"method": method},
        )
    if has_result and not isinstance(payload["result"], dict):
        raise AcsError(
            INVALID_PARAMS, "a wrapped MCP response's result must be an object", data={"method": method}
        )
    if has_error:
        error = payload["error"]
        if (
            not isinstance(error, dict)
            or not isinstance(error.get("code"), int)
            or isinstance(error.get("code"), bool)
            or not isinstance(error.get("message"), str)
        ):
            raise AcsError(
                INVALID_PARAMS,
                "a wrapped MCP response's error must be an object with an integer code and a string message",
                data={"method": method},
            )
    return payload
